msu/systems/serialization/serde_emulator.nut

```diff
@@ -1,45 +1,170 @@
-			throw ::MSU.Exception.InvalidValue(_id);
+			::MSU.requireInt(_int);
+			this.__writeData(_int);
 		}
-		this.Data = [];
-		this.Mod = _mod;
-		this.ID = _id;
-		this.FlagContainer = _flagContainer;
+
+		function writeU32( _int )
+		{
+			::MSU.requireInt(_int);
+			if (_int < 0)
+				throw ::MSU.Exception.InvalidValue(_int);
+			this.__writeData(_int);
+		}
+
+		function writeI16( _int )
+		{
+			::MSU.requireInt(_int);
+			if (_int < -32768 || _int > 32767)
+				throw ::MSU.Exception.InvalidValue(_int);
+			this.__writeData(_int);
+		}
+
+		function writeU16( _int )
+		{
+			::MSU.requireInt(_int);
+			if (_int < 0 || _int > 65535)
+				throw ::MSU.Exception.InvalidValue(_int);
+			this.__writeData(_int);
+		}
+
+		function writeI8( _int )
+		{
+			::MSU.requireInt(_int);
+			if (_int < -128 || _int > 127)
+				throw ::MSU.Exception.InvalidValue(_int);
+			this.__writeData(_int);
+		}
+
+		function writeU8( _int )
+		{
+			::MSU.requireInt(_int);
+			if (_int < 0 || _int > 255)
+				throw ::MSU.Exception.InvalidValue(_int);
+			this.__writeData(_int);
+		}
+
+		function writeF32( _float )
+		{
+			::MSU.requireOneFromTypes(["float", "integer"], _float);
+			this.__writeData(_float);
+		}
+	};
+
+	constructor( _metaDataEmulator, _serializationData = null )
+	{
 		this.MetaData = _metaDataEmulator;
+		if (_serializationData == null)
+			this.resetData();
+		else
+		{
+			::MSU.requireInstanceOf(::MSU.Class.SerializationData, _serializationData);
+			this.SerializationData = _serializationData;
+		}
 	}
 
-	function getEmulatorString()
+	function resetData()
 	{
-		return format("MSU.%s.%s", this.Mod.getID(), this.ID);
+		this.SerializationData = ::MSU.Class.SerializationData();
 	}
 
-	function clearFlags()
+	function getData()
 	{
-		local startString = this.getEmulatorString();
-		this.FlagContainer.remove(startString);
-		for (local i = 0; i < this.Data.len(); ++i)
-		{
-			this.FlagContainer.remove(startString + "." + i);
-		}
+		return this.SerializationData;
 	}
 
 	function getMetaData()
 	{
 		return this.MetaData;
 	}
+
+	function __writeData( _data )
+	{
+		this.SerializationData.push(_data);
+	}
+
+	function __readData( _type )
+	{
+		if (this.SerializationData.len() <= ++this.Idx)
+		{
+			::logError(format("Tried to read data beyond (%i) the length (%i) of the Deserialization Emulator", this.Idx, this.SerializationData.len()));
+			return null;
+		}
+		local data = this.SerializationData.getDataArray()[this.Idx];
+		if (data.getType() != _type)
+		{
+			::logError(format("The type being read %s isn't the same as the type %s stored in the Deserialization Emulator", ::MSU.Serialization.DataType.getKeyForValue(_type), ::MSU.Serialization.DataType.getKeyForValue(this.SerializationData.getDataArray()[this.Idx].getType())));
+			// currently still continuing in case of conversion between integers
+		}
+		return data.getData();
+	}
```

Comment:
Also these should be in `__XXXXFunctions` unless I'm missing something

Reply:
Children have to overwrite these functions so I couldn't find a better way to do it. Open to suggestions.